File: CodeGenDotNet 159059-1372 (a)/VBCode/Chapter 1/HelloWorldAndSimpleClass/Startup.vb
Option Strict On
Option Explicit On

Imports System

Public Structure XSLTParam
   Public Name As String
   Public Value As String
   Public Sub New(ByVal Name As String, ByVal Value As String)
      Me.Name = Name
      Me.Value = Value
   End Sub
End Structure

'! Class Summary: 

Public Class Startup
   Private Const outputDir As String = "..\Test"
   Private Const xsltDir As String = ".."
   Private Const xmlDir As String = ".."
   Private Const tablename As String = "Customers"
   ' Output filenames are hard coded in this simple sample

#Region "Class level declarations"
#End Region

#Region "Constructors"
#End Region

   Public Shared Sub Main()
      ' Be sure output directory exists
      If Not IO.Directory.Exists(outputDir) Then
         IO.Directory.CreateDirectory(outputDir)
      End If

      Test()
      GenerateHelloWorld(outputDir)
      GenerateClassViaBruteForce(outputDir)
      GenerateClassViaCodeDOM(outputDir)
      GenerateClassViaXSLT(outputDir)

   End Sub

   Private Shared Sub GenerateHelloWorld(ByVal outputDir As String)

      ' Generate Hello World using Brute Force
      HelloWorldViaBruteForce.GenerateOutput(IO.Path.Combine(outputDir, "HelloWorldViaBruteForce.vb"))

      ' Generate Hello World in Visual Basic using the CodeDOM
      Dim compileUnit As CodeDom.CodeCompileUnit
      Dim provider As CodeDom.Compiler.CodeDomProvider
      compileUnit = HelloWorldViaCodeDOM.BuildGraph()
      provider = New Microsoft.VisualBasic.VBCodeProvider
      GenerateViaCodeDOM(IO.Path.Combine(outputDir, "HelloWorldViaCodeDOM.vb"), provider, compileUnit)
      ' Use same compile unit to generate C# Hello World
      provider = New Microsoft.CSharp.CSharpCodeProvider
      GenerateViaCodeDOM(IO.Path.Combine(outputDir, "HelloWorldViaCodeDOM.cs"), provider, compileUnit)

      ' Generate Hello World via XSLT Template
      GenerateViaXSLT(IO.Path.Combine(xsltDir, "HelloWorld.xslt"), Nothing, IO.Path.Combine(outputDir, "HelloWorldViaXSLT.vb"))

   End Sub

   Private Shared Sub GenerateClassViaBruteForce(ByVal outputDir As String)
      ' Open Metadata file
      Dim xmlMetaData As New Xml.XmlDocument
      xmlMetaData.Load(IO.Path.Combine(xmlDir, "Metadata.xml"))

      ClassViaBruteForce.GenerateOutput(IO.Path.Combine(outputDir, "ClassCustomersViaBruteForce.vb"), xmlMetaData, "Customers")
      ClassViaBruteForce.GenerateOutput(IO.Path.Combine(outputDir, "ClassOrdersViaBruteForce.vb"), xmlMetaData, "Orders")

   End Sub


   Private Shared Sub GenerateClassViaCodeDOM(ByVal outputDir As String)
      Dim compileUnit As CodeDom.CodeCompileUnit
      Dim provider As CodeDom.Compiler.CodeDomProvider

      ' Open Metadata file
      Dim xmlMetaData As New Xml.XmlDocument
      xmlMetaData.Load(IO.Path.Combine(xmlDir, "Metadata.xml"))

      ' Generate simple class for Cusotmers in Visual Basic using the CodeDOM
      compileUnit = ClassViaCodeDOM.BUildGraph(xmlMetaData, "Customers")
      provider = New Microsoft.VisualBasic.VBCodeProvider
      GenerateViaCodeDOM(IO.Path.Combine(outputDir, "ClassCustomersViaCodeDOM.vb"), provider, compileUnit)

      ' Generate simple class for Orders in Visual Basic using the CodeDOM
      compileUnit = ClassViaCodeDOM.BUildGraph(xmlMetaData, "Orders")
      provider = New Microsoft.VisualBasic.VBCodeProvider
      GenerateViaCodeDOM(IO.Path.Combine(outputDir, "ClassOrdersViaCodeDOM.vb"), provider, compileUnit)
   End Sub

   Private Shared Sub GenerateClassViaXSLT(ByVal outputDir As String)
      ' Open Metadata file
      Dim xmlMetaData As New Xml.XmlDocument
      xmlMetaData.Load(IO.Path.Combine(xmlDir, "Metadata.xml"))

      ' Generate Hello World via XSLT Template
      GenerateViaXSLT(IO.Path.Combine(xsltDir, "Class.xslt"), xmlMetaData, IO.Path.Combine(outputDir, "ClassCustomersViaXSLT.vb"), New XSLTParam("TableName", "Customers"))
      GenerateViaXSLT(IO.Path.Combine(xsltDir, "Class.xslt"), xmlMetaData, IO.Path.Combine(outputDir, "ClassOrdersViaXSLT.vb"), New XSLTParam("TableName", "Orders"))
   End Sub

   Private Shared Sub GenerateViaCodeDOM( _
               ByVal outputFileName As String, _
               ByVal provider As CodeDom.Compiler.CodeDomProvider, _
               ByVal compileunit As CodeDom.CodeCompileUnit)
      Dim gen As CodeDom.Compiler.ICodeGenerator = provider.CreateGenerator()
      Dim tw As CodeDom.Compiler.IndentedTextWriter

      tw = New CodeDom.Compiler.IndentedTextWriter(New IO.StreamWriter(outputFileName, False), "    ")
         gen.GenerateCodeFromCompileUnit(compileunit, tw, New CodeDom.Compiler.CodeGeneratorOptions)
         tw.Flush()
         tw.Close()

   End Sub


   Private Shared Sub GenerateViaXSLT( _
               ByVal xsltFileName As String, _
               ByVal xmlMetaData As Xml.XmlDocument, _
               ByVal outputFile As String, _
               ByVal ParamArray params() As XSLTParam)
      Dim xslt As New Xml.Xsl.XslTransform
      Dim xNav As Xml.XPath.XPathNavigator
      Dim streamWriter As IO.StreamWriter
      Dim args As New Xml.Xsl.XsltArgumentList
      Dim param As XSLTParam

      Try
         If xmlMetaData Is Nothing Then
            xmlMetaData = New Xml.XmlDocument
         End If

         For Each param In params
            args.AddParam(param.Name, "", param.Value)
         Next

         xNav = xmlMetaData.CreateNavigator()
         streamWriter = New IO.StreamWriter(outputFile)

         xslt.Load(xsltFileName)
         xslt.Transform(xNav, args, streamWriter, Nothing)

      Finally
         If Not streamWriter Is Nothing Then
            streamWriter.Flush()
            streamWriter.Close()
         End If
      End Try

   End Sub

   Private Shared Sub test()
      Debug.WriteLine("*" & CodeDom.Compiler.IndentedTextWriter.DefaultTabString & "*")
      Dim tw As New IO.StreamWriter("..\IndentTest")
      Dim itw As New CodeDom.Compiler.IndentedTextWriter(tw)
      itw.WriteLine("Fred")
      itw.Indent = 2
      itw.WriteLine("Fred")
      itw.Indent = 4
      itw.WriteLine("Fred")
      itw.Indent = itw.Indent + 1
      itw.WriteLine("Fred")
      itw.Indent = itw.Indent - 1
      itw.WriteLine("Fred")
      itw.Indent = itw.Indent - 1
      itw.WriteLine("Fred")
      itw.Indent = itw.Indent - 1
      itw.Flush()
      itw.Close()
   End Sub

End Class

File: CodeGenDotNet 159059-1372 (a)/CSharp/Chapter 1/HelloWorldAndSimpleClass/Startup.cs
using System;

public struct XSLTParam
{
	public string Name;
	public string Value;
	public XSLTParam( string Name, string Value )
	{
		this.Name = Name;
		this.Value = Value;
	}
}

/// <summary>
/// 
/// </summary>
public class Startup
{
	private const string outputDir = "..\\..\\Test";
	private const string xsltDir = "..\\..";
	private const string  xmlDir = "..\\..";
	private const string  tablename = "Customers";
	// Output file names are hard coded in this simple example
	
	#region Class level declarations
	#endregion

	#region Constructors
	#endregion

	public static void Main()
	{
		// Be sure output directory exists
		if( !System.IO.Directory.Exists( outputDir ) )
		{
			System.IO.Directory.CreateDirectory( outputDir );
		}

		GenerateHelloWorld(outputDir);
		GenerateClassViaBruteForce(outputDir);
		GenerateClassViaCodeDOM(outputDir);
		GenerateClassViaXSLT(outputDir);
	}

	private static void GenerateHelloWorld( string outputDir )
	{
		// Generate Hello World using Brute Force
		HelloWorldViaBruteForce.GenerateOutput( System.IO.Path.Combine( outputDir, "HelloWorldViaBruteForce.cs" ) );

		// Generate Hello World in Visual Basic using the CodeDOM
		System.CodeDom.CodeCompileUnit compileUnit;
		System.CodeDom.Compiler.CodeDomProvider provider;
		compileUnit = HelloWorldViaCodeDOM.BuildGraph();
		provider = new Microsoft.CSharp.CSharpCodeProvider();
		GenerateViaCodeDOM( System.IO.Path.Combine( outputDir, "HelloWorldViaCodeDOM.cs" ), provider, compileUnit );
		// Use same compile unit to generate VB Hello World
		provider = new Microsoft.VisualBasic.VBCodeProvider();
		GenerateViaCodeDOM( System.IO.Path.Combine( outputDir, "HelloWorldViaCodeDOM.vb" ), provider, compileUnit );

		// Generate Hello World via XSLT Template
		GenerateViaXSLT(	System.IO.Path.Combine( xsltDir, "HelloWorld.xslt" ), 
							null, 
							System.IO.Path.Combine( outputDir, "HelloWorldViaXSLT.cs" ) );
	}

	private static void GenerateClassViaBruteForce( string outputDir )
	{
		// Open Metadata file
		System.Xml.XmlDocument xmlMetaData = new System.Xml.XmlDocument();
		xmlMetaData.Load( System.IO.Path.Combine( xmlDir, "Metadata.xml" ) );

		ClassViaBruteForce.GenerateOutput( System.IO.Path.Combine( outputDir, "ClassCustomersViaBruteForce.cs" ), xmlMetaData, "Customers" );
		ClassViaBruteForce.GenerateOutput( System.IO.Path.Combine( outputDir, "ClassOrdersViaBruteForce.cs" ), xmlMetaData, "Orders" );
	}

	private static void GenerateClassViaCodeDOM( string outputDir )
	{
		System.CodeDom.CodeCompileUnit compileUnit;
		System.CodeDom.Compiler.CodeDomProvider provider;

		// Open Metadata file
		System.Xml.XmlDocument xmlMetaData = new System.Xml.XmlDocument();
		xmlMetaData.Load( System.IO.Path.Combine( xmlDir, "Metadata.xml" ) );

		// Generate simple class for Cusotmers in C# using the CodeDOM
		compileUnit = ClassViaCodeDOM.BuildGraph( xmlMetaData, "Customers" );
		provider = new Microsoft.CSharp.CSharpCodeProvider();
		GenerateViaCodeDOM( System.IO.Path.Combine( outputDir, "ClassCustomersViaCodeDOM.cs" ), provider, compileUnit );
		
		// Generate simple class for Orders in C# using the CodeDOM
		compileUnit = ClassViaCodeDOM.BuildGraph( xmlMetaData, "Orders" );
		GenerateViaCodeDOM( System.IO.Path.Combine( outputDir, "ClassOrdersViaCodeDOM.cs" ), provider, compileUnit );
	}

	private static void GenerateClassViaXSLT( string outputDir )
	{
		// Open Metadata file
		System.Xml.XmlDocument xmlMetaData = new System.Xml.XmlDocument();
		xmlMetaData.Load( System.IO.Path.Combine( xmlDir, "Metadata.xml" ) );

		// Generate Hello World via XSLT Template
		GenerateViaXSLT( System.IO.Path.Combine( xsltDir, "Class.xslt" ), xmlMetaData, System.IO.Path.Combine( outputDir, "ClassCustomersViaXSLT.cs" ), new XSLTParam( "TableName", "Customers" ) );
		GenerateViaXSLT( System.IO.Path.Combine( xsltDir, "Class.xslt" ), xmlMetaData, System.IO.Path.Combine( outputDir, "ClassOrdersViaXSLT.cs" ), new XSLTParam( "TableName", "Orders" ) );
	}

	private static void GenerateViaCodeDOM( string outputFileName, System.CodeDom.Compiler.CodeDomProvider provider, System.CodeDom.CodeCompileUnit compileUnit )
	{
		System.CodeDom.Compiler.ICodeGenerator gen = provider.CreateGenerator();
		System.CodeDom.Compiler.IndentedTextWriter tw;

		tw = new System.CodeDom.Compiler.IndentedTextWriter( new System.IO.StreamWriter( outputFileName, false ), "	" );
		System.CodeDom.Compiler.CodeGeneratorOptions options = new System.CodeDom.Compiler.CodeGeneratorOptions();
		options.BracingStyle = "C";
		gen.GenerateCodeFromCompileUnit( compileUnit, tw, options );
		tw.Flush();
		tw.Close();
	}

	private static void GenerateViaXSLT(	string xsltFileName,
											System.Xml.XmlDocument xmlMetaData,
											string outputFile,
											params XSLTParam[] @params )
	{
		System.Xml.Xsl.XslTransform xslt = new System.Xml.Xsl.XslTransform();
		System.Xml.XPath.XPathNavigator xNav;
		System.IO.StreamWriter streamWriter = null;
		System.Xml.Xsl.XsltArgumentList args = new System.Xml.Xsl.XsltArgumentList();
		
		try
		{
			if( xmlMetaData == null )
			{
				xmlMetaData = new System.Xml.XmlDocument();
			}
			foreach( XSLTParam param in @params )
			{
				args.AddParam( param.Name, "", param.Value );
			}

			xNav = xmlMetaData.CreateNavigator();
			streamWriter = new System.IO.StreamWriter( outputFile );

			xslt.Load( xsltFileName );
			xslt.Transform( xNav, args, streamWriter, null );
		}
		finally
		{
			if( streamWriter != null )
			{
				streamWriter.Flush();
				streamWriter.Close();
			}
		}
	}
}